File: lua/vimade/style/component.lua
local M = {}

local CONDITION = require('vimade.style.value.condition')

-- components do nothing except act as a named condition allowing users to override things based on concept
-- @param componentName: string
-- @param config = {
--  style = {Fade(0.4)} # style to run on all names that aren included
--}
M.Component = function(name, config)
  local result = {}
  local _condition = config.condition or CONDITION.INACTIVE
  result.name = name
  result.tick = config.tick
  result.attach = function (win)
    local condition = _condition
    local children = {}
    for i, s in ipairs(config.style) do
      table.insert(children, s.attach(win))
    end
    local style = {}
    style.win = win
    style._condition = _condition
    style._animating = false
    style.before = function (win, state)
      if type(_condition) == 'function' then
        condition = _condition(style, state)
      end
      if condition == false then
        return
      end
      for i, s in ipairs(children) do
        s.before(win, state)
      end
    end
    style.key = function (win, state)
      -- components don't need their own keys since they are just proxies to their children
      if condition == false then
        return ''
      end
      local key = ''
      local s_active = 0
      for j, s in ipairs(children) do
        if j ~= 0 then
          key = key .. ','
        end
        local s_key = s.key(win, state)
        key = key .. s_key
        if string.len(s_key) > 0 then
          s_active = s_active + 1
        end
      end
      if s_active == 0 then
        return ''
      end
      return key
    end
    style.modify = function (highlights, to_hl)
      if condition == false then
        return
      end
      for i, s in ipairs(children) do
        s.modify(highlights, to_hl)
      end
    end
    return style
  end
  -- value is not exposed here, no use-case currently
  return result
end
return M

File: lib/vimade/style/component.py
from vimade.style.value import condition as CONDITION

class Component():
  def __init__(parent, name, **kwargs):
    _condition = kwargs.get('condition', CONDITION.INACTIVE)
    parent.tick = kwargs.get('tick')
    parent.name = name
    class __Component():
      def __init__(self, win, state):
        self.win = win
        self._condition = _condition
        self._animating = False
        self.style = [s.attach(win, state) for s in kwargs.get('style', [])]
      def before(self, win, state):
        self.condition = _condition(self, state) if callable(_condition) else _condition
        if self.condition == False:
          return
        for s in self.style:
          s.before(win, state)
      def key(self, win, state):
        # components don't need their own keys since they are just proxies to their children
        if self.condition == False:
          return ''
        style_keys = [s.key(win, state) for s in self.style]
        return ','.join([s for s in style_keys if s])
      def modify(self, highlights, to_hl):
        if self.condition == False:
          return
        for s in self.style:
          s.modify(highlights, to_hl)
    parent.attach = __Component
